execute/exectypes/type_optimizer.go

```diff
@@ -0,0 +1,209 @@
+				}
+				obs.Messages[chain][seqNum] = cciptypes.Message{}
+				obs.TokenData[chain][seqNum] = NewMessageTokenData()
+				// Subtract the removed message and token size
+				encodedObsSize -= messageAndTokenDataEncodedSizes[chain][seqNum]
+				// Add empty message and token encoded size
+				encodedObsSize += op.emptyEncodedSizes.MessageAndTokenData
+				seqNum++
+				if encodedObsSize <= op.maxEncodedSize {
+					return obs, nil
+				}
+			}
+
+			var sizeDelta int
+			// Reaching here means that all messages in the report are truncated, truncate the last commit
+			obs, sizeDelta = op.truncateLastCommit(obs, chain)
+
+			encodedObsSize += sizeDelta
+
+			if len(obs.CommitReports[chain]) == 0 {
+				// If the last commit report was truncated, truncate the chain
+				obs = op.truncateChain(obs, chain)
+			}
+
+			if encodedObsSize <= op.maxEncodedSize {
+				return obs, nil
+			}
+			chains = maps.Keys(obs.CommitReports)
+		}
+		// Truncated all chains. Return obs as is. (it has other data like contract discovery)
+		if len(obs.CommitReports) == 0 {
+			return obs, nil
+		}
+		// Encoding again after doing a full iteration on all chains and removing messages/commits.
+		// That is because using encoded sizes is not 100% accurate and there are some missing bytes in the calculation.
+		encodedObs, err = obs.Encode()
+		if err != nil {
+			return Observation{}, nil
+		}
+		encodedObsSize = len(encodedObs)
+	}
+
+	return obs, nil
+}
+
+// truncateLastCommit removes the last commit from the observation.
+// errors if there are no commits to truncate.
+func (op ObservationOptimizer) truncateLastCommit(
+	obs Observation,
+	chain cciptypes.ChainSelector,
+) (Observation, int) {
+	observation := obs
+	sizeDelta := 0
+	commits := observation.CommitReports[chain]
+	if len(commits) == 0 {
+		return observation, sizeDelta
+	}
+	lastCommit := commits[len(commits)-1]
+	// Remove the last commit from the list.
+	commits = commits[:len(commits)-1]
+	observation.CommitReports[chain] = commits
+	// Remove from the encoded size.
+	sizeDelta = sizeDelta - op.emptyEncodedSizes.CommitData - 4 // brackets, and commas
+	for seqNum, msg := range observation.Messages[chain] {
+		if lastCommit.SequenceNumberRange.Contains(seqNum) {
+			// Remove the message from the observation.
+			delete(observation.Messages[chain], seqNum)
+			// Remove the token data from the observation.
+			delete(observation.TokenData[chain], seqNum)
+			//delete(observation.Hashes[chain], seqNum)
+			// Remove the encoded size of the message and token data.
+			sizeDelta -= op.emptyEncodedSizes.MessageAndTokenData
+			sizeDelta -= 2 * op.emptyEncodedSizes.SeqNumMap
+			sizeDelta -= 4 // for brackets and commas
+			// Remove costly messages
+			for i, costlyMessage := range observation.CostlyMessages {
+				if costlyMessage == msg.Header.MessageID {
+					observation.CostlyMessages = append(observation.CostlyMessages[:i], observation.CostlyMessages[i+1:]...)
+				}
+			}
```

Comment:
Have you checked if we have a sort of race condition here? We are modifying the array while looping through it. Might be working well but feels like we doing something wrong. Same for `truncateChain`

Reply:
I think this would have been caught in CI tests which runs a lot of times and detect races. I assume that means it's working fine. WDYT @winder ?

Reply:
Not sure about a race, but @0xnogo is right that something is wrong. Two problems I see:
1. It only works if you're only removing one item. After removing the first, `i` would be offset to an incorrect value.
2. You cannot remove the final value, there would be an out of bounds panic because `i+1` is the index after the last element.

I usually copy them into a new slice to avoid dealing with array details:
```
			var newCostlyMessages []Message
			for _, costlyMessage := range observation.CostlyMessages {
				if costlyMessage != msg.Header.MessageID {
					newCostlyMessages = append(newCostlyMessages, costlyMessage)
				}
			}
			observation.CostlyMessages = newCostlyMessages
```

You could also copy them in place:
```
			i := 0
			for _, costlyMessage := range observation.CostlyMessages {
				if costlyMessage != msg.Header.MessageID {
					observation.CostlyMessages[i] = costlyMessage
					i++
				}
			}
			observation.CostlyMessages = observation.CostlyMessages[:i]
```

Reply:
I added a utility function `RemoveIthElement` and using it now 👍